text/001x-battery_improvements.md

```diff
@@ -0,0 +1,293 @@
+- New `battery_status` (`MAV_BATTERY_STATUS_FLAGS`) replaces [`fault_bitmask`](https://mavlink.io/en/messages/common.html#MAV_BATTERY_FAULT), [`charge_state`](https://mavlink.io/en/messages/common.html#MAV_BATTERY_CHARGE_STATE) and [`mode`](https://mavlink.io/en/messages/common.html#MAV_BATTERY_MODE)
+  ```xml
+  <field type="uint8_t" name="charge_state" enum="MAV_BATTERY_CHARGE_STATE">State for extent of discharge, provided by autopilot for warning or external reactions</field>
+  <field type="uint8_t" name="mode" enum="MAV_BATTERY_MODE">Battery mode. Default (0) is that battery mode reporting is not supported or battery is in normal-use mode.</field>
+  <field type="uint32_t" name="fault_bitmask" display="bitmask" enum="MAV_BATTERY_FAULT">Fault/health/ready-to-use indications.</field>
+  ```
+  - The state is almost all inferred from the battery remaining. Traditionally the GCS sets appropriate levels for warning - the battery has a less accurate view on appropriate warning levels for a particular system, making this redundant. The exception is charging status.
+  - The mode and fault are reasonable, but make more sense as general status indication.
+  - The status loses some faults which make no sense (e.g. `MAV_BATTERY_FAULT_SPIKES`) or which can now be inferred (like deep discharge).
+
+
+#### Questions
+
+- Do we need all the other fields?
+- Are there any other fields missing?
+- Should we have `MAV_BATTERY_FAULT` for critical level "just before deep discharge"? The battery can know this, and it might prevent over discharge.
+  ```xml
+      <entry value="1024" name="BATTERY_FAULT_CRITICAL_LEVEL">
+        <description>Battery is at critically low level. Undamaged, but not ready to use.</description>
+      </entry>
+  ```
+- `MAV_BATTERY_STATUS_FLAGS`
+  - Should this include `MAV_BATTERY_STATUS_FLAGS_IN_USE` (connected, drawing significant power to fly). Not clear when you would need this or how you would use it. Or precisely what would count as "in use".
+
+## Battery Cell Voltages
+
+The proposed battery message is:
+
+```xml
+    <message id="371" name="BATTERY_CELL_VOLTAGES">
+      <description>Battery cell voltages.
+        This message is provided primarily for cell fault debugging.
+        For batteries with more than 12 cells the message should be sent multiple times, iterating the index value.
+        It should be streamed at very low rate (less than once a minute) or streamed only on request.</description>
+      <field type="uint16_t" name="id" instance="true">Battery ID</field>
+      <field type="uint16_t" name="index">Cell index (0 by default). This can be iterated for batteries with more than 12 cells.</field>
+      <field type="uint16_t[126]" name="voltages" units="mV" invalid="[0]">Battery voltage of up to 126 cells at current index. Cells above the valid cell count for this battery should be set to 0.</field>
```

Comment:
was this supposed to be [12] not [126] ?
we also need a count to be able to tell how many are valid, and count need to be uint16_t to get the right packing and zero truncation

Reply:
i'd prefer to match the proposed DrroneCAN one, and make length 24, with index of the first cell filled

Reply:
@tridge 

This is supposed to be 126. Thought was to reduce overheads of sending multiple messages and the hassle of re-assembling them until you get to 126 cells.

The array will be ordered last and any unused cells are zero-filled, so this will truncate on the number of cells used. No  count needed". If we set this to 24 cells the only difference is that for more than 24 cells you'll have to reconstruct the full cell list earlier. What's the value in that other than being able to extend the message later? 

Note - other than the length the UAVCAN message looks the same: https://github.com/dronecan/DSDL/pull/7/files

Reply:
I think the purpose was that by mirroring both messages as close as possible there would be fewer implementation details?

Reply:
Fair enough. I'd like confirmation though, because that's not what it sounds like. I'm fairly confident of my understanding of MAVLink, but then this is Tridge, who is very rarely wrong.

Reply:
if its supposed to be 12 then why does it say "This can be iterated for batteries with more than 12 cells"

Reply:
@tridge tridge Thanks, that's a typo. Fixed to 126.

But you also said 
- we also need a count to be able to tell how many are valid, and count need to be uint16_t to get the right packing and zero truncation
- i'd prefer to match the proposed DrroneCAN one, and make length 24, with index of the first cell filled

So I thought there might be an actual problem with the design of the message being 126. I can force this to 24 to match DroneCan if you feel strongly that is somehow better. The symmetry appeals, it's just I can't see a justification for it technically.

Reply:
I consider the is message "on hold" (though non blocking on the rest of it). From @tridge:

> Needed:
> - a way to cope with high voltage batteries for total voltage 
> - a way to know if a cell is failing
>
> we can get that with float total_voltage and with highest/lowest cell voltage
> vastly easier to sound an alarm on GCS too we'd just have a percentage deviation threshold

I've requested a PR to this message.

Reply:
see my proposal for a max/min cell voltage, much saner

Reply:
https://github.com/mavlink/rfcs/pull/19#issuecomment-1275544121